README.md
# padachone
Production Env of padachone.com


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Map showing mosque locations
- Custom Search Results
- Current location data
- Clean Lab

- firebase : https://sebhastian.com/react-firestore
- read on https://opensource.guide/
- newsletters in lab
- to do : flash of all importatn users
- to do : custom trace and metrics on app using firebase performance monitoring
- to do : schedule emails using Bull
- to do : use nice font
- to do  : change git email
- to do : avoid security leak by hiding credentials
- to do : optimize bundle, cdn
- to do : admin dashboard
- to do :log rocket
- to do : testing & end to end testing using puppeteer
- to do : skeletons using material ui
- to do : make text and bg color dynamic with prismic
- to do : JWT for api & protect routes
### CHECKLIST
- REMOVE DEV while going PROD (package-lock, package.json)
- Add date for changelog release
### NOTES
- Always look at handleexit when a new ls or ss is added
- loggin mechanism with prefixes
- simulation of crash is available at Setup.js
- Use `GR` for inspecting graphQL data logs

## [3.2.1] - 2020-02-29
### Added
- Youtube media playback to Umrah video
- New to 1
- npm audit fix


## [3.2.0] - 2019-11-09
### Added
- Media Playback at homescreen (Prayertimes screen) with fullscreen paper
- Hook up with Prismic

### Changed
- Modal actions buttons are aligned to left in Media display

## [3.1.0] - 2019-10-28
### Added
- Code Splitting with lazy-suspense for Travel
- Log for crash report

### Changed
- Replaced APOD word to the apod title
- Replaced hardcoded place example with visitors city & postal code

## [3.0.1] - 2019-10-14
### Added
- username display on click of u in Know Your with modal display
- logs for Prayer times & dismiss functionality
- log detected location in Traveller onboard & hijiri from Travel

### Changed
- username & token would now persist even if you start over from set me up screen
- Apod is updated when American time changes
- Message prefix to  logged in log message
- Apod is now dependent only on American date as user's date is not being passed as current date anymore
- Ignore hosts for loggers

## [3.0.0] - 2019-10-12
### Added
- FCM
- Handled firebase push api check for browser (meant for safari)
- FCM token to firebase
- Reminder : Hook up notification & easy cron
- Dev Banner
- Hook up with timezone
- Disable reminder on safari
- Test Alert modal window
- In App Notification Display Implementation
- Metadata expansion panel for prayers for showing hijri details
- Accordion component
- hid alarm off for ios
- api for sending FCM notification and hooked it up with test reminder
- api call for setting scheduling
- Color block for hijri meta data
- Twilio Integration for logging
- Whatsapp logger custom hook & implemented in all important events
- A unique username generated for all users in addition to hooking up with logger and firebase
- EMojis in whatsapp logger
- Hijri to Travel screen
- Apod for Landing page
- Conditional fetching of apod based on dynamic source - prismic or application
- Apod details screen with redirection
- Moving bg for apod
- Hooked whatsapp logging to apod and learn more
- Dynamically (prismic) overridden text and bg color for landing bg for apod sake

### Changed
- gitignore
- Appended GCM id in manifest.json
- Hotfix implementation for hiding blank timer display 
- Newsletter component uses shared Accordion component
- Prayer time display rendering style for home screen
- Hot fix to rectify blank timer display
- display either timezone or reminder
- restrict reminder for just prayers
- changed 1 min test reminder to 2 min reminder
- Max stackbar is limited to 1 in infomessage of travel
- Got rid of redundant hijri month display from travel 
- Made infomessage in travel top center
- Autohide duration of infomessage to 3s
- Customized hijri for the travel
- Also deleted testreminder & apod ls while logging out
- Example address of Set me up process

## [2.1.3] - 2019-09-04
### Changed
- Landing bg image, bg color & font color are prismic bound now
- Site description is dynamic
- Site footer is prismic bound

## [2.1.2] - 2019-08-27
### Changed
- Country is now pre populated even after hard reset
- Bg image of grid menus has been made prismic bound
- Title bg image is also prismic bound now
- Prayertime boy is prismic bound
- emojis are also prismic tied
- logo is dynamic but with a fallback option
- resolved e2e bug by deleting unused webapp.test file

## [2.1.1] - 2019-08-..
### Added
- puppeteer as a dev dependency for end to end testing
- code coverage using page-coverage dev dependency
### Changed
- Modified const to let in Travel times Asis/Calcutta hot fix

## [2.1.0] - 2019-08-12
### Added
- Apollo client & graphql for prismic
- Materialui LAB for skeleton
- Newsletters list in Lab
- added graphql document names as constants
- Bgm sets in when prayer time boy appears
- Made timeout display dynamic with Prismic
### Changed
- Message display to accomodate image as well
- Removed prismic helper at hooks side

## [2.0.4] - 2019-08-06
### Added
- Set Display timer to false after the last prayer of the day as part of fixing the same defect

## [2.0.3] - 2019-08-06
### Changed
- Put the hot fix (Asia/Calcutta timezone) in traveller onboard pages

## [2.0.2] - 2019-07-27
### Changed
- Null check for location data in Traveller onboard

## [2.0.1] - 2019-07-21
### Changed
- Put a hot fix for Asia/Calcutta timezone

## [2.0.0] - 2019-07-14
### Added
- Fine tuning option for setting calculation methods and school preferences
- Integration of Fine tune with side menus, Travel & homepage
- Caching of calculation methods
- Address input acceptance
- Facebook business chat plugin integration
- New landing page with grid menus

src/components/App/App.js
import React, { useState, useEffect } from "react";
import "./App.css";
import CssBaseline from "@material-ui/core/CssBaseline";
// import 'typeface-roboto';
import Layout from "../Layout";
import { createMuiTheme } from "@material-ui/core/styles";
import { ThemeProvider } from "@material-ui/styles";
import lightBlue from "@material-ui/core/colors/lightBlue";
import Setup from "../Setup/Setup";
import stepperData from "../Setup/setup-stepper-data.json";
import { getPDdata, getDateTimeOf } from "../../utils";
import ErrorBoundary from "../Error/ErrorBoundary";
import CookieConsent from "react-cookie-consent";
import Messages from "../Messages";
import SpecialDay from "../Messages/SpecialDay";
import Zoom from "@material-ui/core/Zoom";
import { UserContext } from "../../store/context/userContext";
import Travel from "../Travel";
import Subscribe from "../Subscribe";
import Finetune from "../Finetune";
import Lab from "../Lab";
import { FT_PRAYER, PRISMIC_MSG_BROADCAST_DOC } from "../../utils/constants";
import LandingPage from "./LandingPage";
import ConfirmAction from "../ConfirmAction";
import {
  useRenderCounts,
  useVisitorDetails,
  useMessageBroadcast
} from "../../hooks/api-hooks";
import Newsletters from "../Newsletters";
import Notify from "../Notification/Notify";
import Banner from "../TopBar";
import { messaging } from "../../config/firebase";
import Apod from "../Apod";
import Media from "../Media";
const theme = createMuiTheme({
  palette: {
    primary: lightBlue,
    secondary: {
      main: "#fff"
    },
    text: {
      // primary: "#000000",
      // secondary: "#ffffff"
    }
  },
  typography: {
    // color: '#fff'
  }
});
function App() {
  useRenderCounts("App.js");
  const [landingpage] = useState("Setup");
  // Web worker part
  const [worker, setWorker] = useState({});
  const [workerData, setWorkerData] = useState({});
  // Web worker part ends here
  const [msgbroadcast] = useMessageBroadcast();
  // Global State
  const [notification, setNotification] = useState({});
  const [tz, setTz] = useState("");
  const [page, setPage] = useState(landingpage);
  console.log("%c PAGE" + page, "font-size:40px;");
  const [iamin, setIamin] = useState(false);
  const [modal, setModal] = useState({ show: false, name: "" });
  // Global State ends here
  const [prevScrollpos, setprevScrollpos] = useState(window.pageYOffset);
  const [display, setdisplay] = useState(true);
  const [forceTrigger, setForceTrigger] = useState({ target: "" });

  const handleForceTrigger = ({ target, method, school }) => {
    setForceTrigger(() => {
      setState({
        ...state,
        method: parseInt(method),
        school: parseInt(school)
      });
      return { target: FT_PRAYER };
    });
  };

  useEffect(() => {
    if (forceTrigger.target === FT_PRAYER) {
      setState({
        ...state,
        method: localStorage.getItem("padachone:method")
          ? parseInt(localStorage.getItem("padachone:method"))
          : 8,
        school: localStorage.getItem("padachone:school")
          ? parseInt(localStorage.getItem("padachone:school"))
          : 0
      });
    }
  }, [forceTrigger]);

  const hideHdrFtr = () => {
    let currentScrollPos = window.pageYOffset;
    if (prevScrollpos >= currentScrollPos) {
      // document.getElementById("navbar").style.top = "0";
      // document.querySelector('.MuiToolbar-gutters').style.display = "flex";
      if (document.querySelector(".padachone-ftr") !== null) {
        document.querySelector(".padachone-ftr").style.visibility = "visible";
      }
    } else {
      // document.getElementById("navbar").style.top = "-50px";
      // document.querySelector('.MuiToolbar-gutters').style.display = "none";
      if (document.querySelector(".padachone-ftr") !== null) {
        document.querySelector(".padachone-ftr").style.visibility = "hidden";
      }
    }
    setprevScrollpos(currentScrollPos);
  };

  useEffect(() => {
    // Web Worker
    const padachoneWorker = new Worker("padachone.worker.js");
    setWorker(padachoneWorker);
    padachoneWorker.addEventListener("message", e => {
      console.log(
        `WORKER: You have a message from worker ${JSON.stringify(e.data.msg)}`
      );
      setWorkerData({ worker_data: e.data });
    });
    // FCM
    // requestNotify(visitor);
    if (messaging) {
      navigator.serviceWorker.addEventListener("message", message => {
        setNotification(message.data);
        handleNav("setNotifymodal");
        console.log("MSG : ", message);
      });
    }
    window.addEventListener("scroll", hideHdrFtr);
    return () => {
      window.removeEventListener("scroll", hideHdrFtr);
    };
  }, []);
  const [state, setState] = useState({
    finished: false,
    pdtodaysDate: getPDdata()
      .split(" ")
      .join(""),
    americanDate: getDateTimeOf("America/New_York")
      .split(",")[0]
      .split(" ")
      .join(""),
    place: localStorage.getItem("padachone:place"),
    country: localStorage.getItem("padachone:country"),
    region: localStorage.getItem("padachone:region"),
    method: localStorage.getItem("padachone:method")
      ? parseInt(localStorage.getItem("padachone:method"))
      : 8,
    school: localStorage.getItem("padachone:school")
      ? parseInt(localStorage.getItem("padachone:school"))
      : 0
  });

  const {
    finished,
    country,
    region,
    pdtodaysDate,
    prayerdata,
    americanDate,
    place,
    method,
    school
  } = state;
  const visitor = useVisitorDetails(pdtodaysDate);
  const handlefinished = obj => {
    const { country, region, finished, place } = obj;
    setState({ ...state, finished, country, region: region, place: place });
    if (finished) {
      setPage(() => {
        setIamin(true);
        return "Home";
      });
    } else {
      setPage(() => {
        setIamin(false);
        return landingpage;
      });
    }
  };

  const handleNav = page => {
    if (page === "callfunc") {
      handleExit();
    } else if (page === "setmodal") {
      setModal({ show: true, name: "Subscribe" });
    } else if (page === "setNotifymodal") {
      setModal({ show: true, name: "Notification" });
    } else if (page === "setFTmodal") {
      setModal({ show: true, name: "Finetune" });
    } else if (page === "setMediamodal") {
      setModal({ show: true, name: "Media" });
    } else if (page === "reset") {
      setModal({
        show: true,
        name: "ConfirmAction",
        message: `<p style={{margin:0}}>This will reset all your settings which include Country, region & address Selection in addition to wiping out Fine tune preferences if any. Would you like to proceed ?</p>`,
        handlePrimary: cb => {
          localStorage.clear();
          sessionStorage.clear();
          cb();
          return window.location.reload();
        },
        handleSecondary: cb => {
          cb();
          return;
        },
        modalconfig: {
          description: "",
          title: "Hard Reset Everything",
          primaryButton: "Yes",
          secondaryButton: "No",
          error: false,
          loading: false
        }
      });
    } else {
      setPage(page);
    }
  };

  const handleExit = () => {
    Object.keys(localStorage).map(key => {
      if (
        key !== "padachone:place" &&
        key !== "padachone:country" &&
        key !== "padachone:region" &&
        key !== "padachone:method" &&
        key !== "padachone:school" &&
        key !== `padachone_FT-${FT_PRAYER}` &&
        key !== `padachone_apod:${americanDate}` &&
        key !== "padachone_testreminder" &&
        key !== "padachone_username" &&
        key !== "padachone_token"
      ) {
        localStorage.removeItem(key);
      }
    });
    handlefinished({
      country: localStorage.getItem("padachone:country"),
      region: localStorage.getItem("padachone:region"),
      place: localStorage.getItem("padachone:place"),
      method: localStorage.getItem("padachone:method"),
      school: localStorage.getItem("padachone:school"),
      finished: false
    });
  };

  const [msg, setMsg] = useState([false, ""]);

  useEffect(() => {
    // Logic for displaying Messages
    localStorage.removeItem("padachone_msg");
    localStorage.removeItem("padachone_msg1");
    localStorage.removeItem("padachone_msg2");
    localStorage.removeItem("padachone_msg3");
    localStorage.removeItem("padachone_msg4");
    localStorage.removeItem("padachone_msg5");
    localStorage.removeItem("padachone_msg6");
    localStorage.removeItem("padachone_msg7");
    if (msgbroadcast) {
      // const message = `Chat with us and pass in your feedback/comments. `;
      setMsg(() => {
        // localStorage.setItem('padachone_msg9', message)
        return [
          true,
          msgbroadcast.data[PRISMIC_MSG_BROADCAST_DOC].edges[0].node
        ];
      });
    }
    // Logic for displaying Messages end here
    const padachon_lsfind = Object.keys(localStorage).filter(
      key =>
        key.startsWith("padachone:") &&
        key !== "padachone:region" &&
        key !== "padachone:country" &&
        key !== "padachone:place" &&
        key !== "padachone:method" &&
        key !== "padachone:school"
    );
    if (padachon_lsfind.length) {
      setState({ ...state, finished: true });
      setPage(() => {
        setIamin(true);
        return "Home";
      });
    }
  }, [msgbroadcast]);
  return (
    <ThemeProvider theme={theme}>
      <div className="App">
        <CssBaseline />
        <UserContext.Provider
          value={{
            tz: tz,
            setTz: setTz,
            page: page,
            setPage: setPage,
            iamin: iamin,
            handleExit: handleExit,
            modal: modal,
            setModal: setModal,
            forceTrigger: forceTrigger,
            setForceTrigger: setForceTrigger,
            handleNav: handleNav,
            visitor: visitor,
            cmsContents: msgbroadcast,
            notification: notification,
            setNotification: setNotification,
            worker: worker,
            workerData: workerData,
            setWorkerData: setWorkerData
          }}
        >
          <ErrorBoundary>
            <CookieConsent
              location="bottom"
              style={{ background: "#29b6f6", marginBottom: "30px" }}
              buttonStyle={{ borderRadius: "10px" }}
            >
              This website uses cookies to enhance the user experience.
            </CookieConsent>
            <Zoom in={true}>
              <SpecialDay display={display} setdisplay={setdisplay} />
            </Zoom>
            {window.location.hostname === "dev.padachone.com" && <Banner />}
            {page === "Newsletters" && <Newsletters />}
            {!finished && page === landingpage && (
              <LandingPage
                finished={locationstate => handlefinished(locationstate)}
                country={country}
                region={region}
                place={place}
              />
            )}
            {msg[0] && <Messages msg={msg[1]} />}
            {page === "Travel" && <Travel method={method} school={school} />}
            {page === "Lab" && (
              <Lab
                timings={{
                  Asr: "18:08",
                  Dhuhr: "13:43",
                  Fajr: "02:59",
                  Imsak: "02:49",
                  Isha: "23:36",
                  Maghrib: "22:06",
                  Midnight: "01:43",
                  Sunrise: "05:20",
                  Sunset: "22:06"
                }}
              />
            )}
            {page === "SetMeup" && (
              <Setup
                setupdata={stepperData}
                finished={locationstate => handlefinished(locationstate)}
                country={
                  country
                    ? country
                    : visitor.country_name
                    ? visitor.country_name
                    : ""
                }
                region={region}
                place={place}
                referrer={landingpage}
              />
            )}
            {page === "Apod" && (
              <Apod pdate={americanDate} referrer={landingpage} />
            )}
            {finished && page === "Home" && (
              <Layout
                country={country}
                region={region}
                pdate={pdtodaysDate}
                place={place}
                method={method}
                school={school}
                startup={resetstate => handlefinished(resetstate)}
              />
            )}
            {modal.show && modal.name === "Subscribe" && (
              <Subscribe modal={modal} setModal={setModal} />
            )}
            {modal.show && modal.name === "Finetune" && (
              <Finetune
                modal={modal}
                setModal={setModal}
                method={method}
                school={school}
                handleForceTrigger={obj => handleForceTrigger(obj)}
              />
            )}
            {modal.show && modal.name === "ConfirmAction" && (
              <ConfirmAction
                modal={modal}
                setModal={setModal}
                message={modal.message}
                handlePrimary={modal.handlePrimary}
                handleSecondary={modal.handleSecondary}
                modalconfig={modal.modalconfig}
              />
            )}

            {modal.show && modal.name === "Notification" && (
              <Notify
                modal={modal}
                setModal={setModal}
                method={method}
                school={school}
                handleForceTrigger={obj => handleForceTrigger(obj)}
              />
            )}

            {modal.show && modal.name === "Media" && (
              <Media modal={modal} setModal={setModal} />
            )}

            {/* <FbChat /> */}
          </ErrorBoundary>
        </UserContext.Provider>
      </div>
    </ThemeProvider>
  );
}

export default App;
